### PR TITLE
Snap packaging support.

### PR DESCRIPTION
**Pull request by wwwtyro**
This PR adds support for snap packages, fixing #44, for the following kubernetes components:

- kubefed
- kubeadm
- kubectl
- kube-apiserver
- kube-controller-manager
- kube-scheduler
- kubelet
- kube-proxy

These follow the same process for deb and RPM by including the appropriate component from the k8s.io build.

### CODE REVIEW

**Comment by luxas** [comment]
Thank you for this patch! Adding snap packages to Kubernetes is just great!

I'll take deeper look into this after KubeCon, but I think we should only build packages for kubelet, kubeadm and kubectl. Also we need a package for the CNI binaries.

There should also be snap packages for all available architectures (see the deb process for more info how to easily do that).

cc @mikedanese @jbeda

**Comment by castrojo** [comment]
cc @marcoceppi

**Comment by marcoceppi** [comment]
Not everyone runs those as containers, many on-premise users and cloud users in the field run those on the "metal" as services. There's no harm in including them - it's an option.

**Comment by jbeda** [comment]
Assigning to @marcoceppi at the request of @castrojo.

**Comment by chuckbutler** [comment]
As this is a known issue moving forward, that we only have AMD64 bins in the snap channels today we have additional incoming feedback requesting PPC64EL.  Linking these two issues together so we can continue to track as the status evolves

https://github.com/juju-solutions/bundle-canonical-kubernetes/issues/268

**Comment by marcoceppi** [comment]
I agree, there is a path forward for multi arch builds of snaps, @ryebot @cynerva is that included in this PR?

**Comment by wwwtyro** [comment]
@marcoceppi It is not; it's here as a PR against this PR branch: https://github.com/juju-solutions/release/pull/5

It's currently blocked by the issues listed there. I think you indicated you were going to think about how to fix those; could be wrong.

**Comment by tvansteenburgh** [comment]
What's blocking this from being merged, anything?

**Comment by marcoceppi** [comment]
I think this is ready for review again? Is that correct @wwwtyro @Cynerva @chuckbutler ?

**Comment by wwwtyro** [comment]
@tvansteenburgh @marcoceppi Last I touched this, we hadn't quite finished the cross-platform snap support, which is what I think was blocking this.

**Comment by tvansteenburgh** [comment]
@wwwtyro Can you summarize what's left to do?

**Comment by wwwtyro** [comment]
@tvansteenburgh I think we're mostly blocked by this and need it fixed or a workaround found: https://bugs.launchpad.net/snapcraft/+bug/1686481

**Comment by dawidmalina** [comment]
Any updates on this? Maybe we could postpone multiarch suport and go with what we already have (more agile - step by step) instead blocking this pull request?

**Comment by luxas** [comment]
What's blocking multiarch support for the snaps?

**Comment by NickZ** [comment]
@luxas it appears to be an upstream bug; snapcraft will not build packages for anything other than the current architecture, even if the architecture is listed in the snapcraft.yaml.

https://bugs.launchpad.net/snapcraft/+bug/1686481?comments=all

Unfortunately, I don't think there's much to be done for multiarch support until upstream fixes this.

**Comment by tvansteenburgh** [comment]
Spoke to kyrofa from the snapcraft team; he requested that we retest using snapcraft 2.34 (it's in artful, or -proposed for zesty and xenial).

**Comment by wwwtyro** [comment]
Just tested against 2.34 - still not working, but it is getting attention.

**Comment by NickZ** [comment]
They assigned someone to the bug, and he says that the current master of snapcraft fixes the problem: https://bugs.launchpad.net/snapcraft/+bug/1686481/comments/12

**Comment by tvansteenburgh** [comment]
Update: we tried this again today using snapcraft from master and unfortunately there are still problems. https://bugs.launchpad.net/snapcraft/+bug/1686481/comments/13

**Comment by NickZ** [comment]
Kyle has updated with a possible fix here: https://bugs.launchpad.net/snapcraft/+bug/1686481/comments/21

**Comment by dims** [comment]
@luxas is this still relevant?

**Comment by luxas** [comment]
Not for me personally at least. I think we're moving towards containers for the control plane & debs/rpms for the kubelet, kubeadm and kubectl, built automatically with bazel

**Comment by calebamiles** [comment]
/hold

where/how snaps should be built and hosted is probably a larger discussion than a single PR, but thanks a lot for the work!
